### PR TITLE
Expose the event object to plugin methods

### LINKED ISSUE #1639: Plugins should have access to the event that prompted the fetch/cache
In some cases plugins may need access to properties of the event object (that triggered the action), but currently they have no way to get at it.

This was the case with the [changes needed](https://github.com/GoogleChrome/workbox/commit/939ba11bb10af3c852a33182894bb7de9ea3b5c7#diff-c0a445b943b01a072d6dce68bef335b3) for the navigation preload package, and is also needed for some changes I'm wanting to make to fix https://github.com/GoogleChrome/workbox/issues/1399.

Rather than passing one-off data as needed, I propose we update [`fetchWrapper.fetch()`](https://github.com/GoogleChrome/workbox/blob/v3.5.0/packages/workbox-core/_private/fetchWrapper.mjs#L39-L42), [`cacheWrapper.put()`](https://github.com/GoogleChrome/workbox/blob/ce7b4d974d44d3a52766b88b5836a1d3c4d64036/packages/workbox-core/_private/cacheWrapper.mjs#L40), and [`cacheWrapper.match()`](https://github.com/GoogleChrome/workbox/blob/v3.5.0/packages/workbox-core/_private/cacheWrapper.mjs#L118) to take named arguments instead of positional arguments, and we can then pass these methods the underlying event whenever it's available (in some cases it won't be). In turn, when plugin lifecycle methods are invoked, they'll be passed the event object if it's available.

### PR DESCRIPTION
**Pull request by philipwalton**
Fixes #1639. 

This PR implements the proposal in #1639 by updating the fetch and catch wrappers as well as the strategies (and other code) that calls them. It also removes the `preloadResponse` option passed to the fetch wrapper, since that can now be found on the fetch event (@jeffposnick can you confirm that's okay?).

In addition, I noticed we're being inconsistent in how we refer to restructured parameters in our JSDoc comments (`options` vs `input`), so, since I was already updating some, I decided to change all instances of `input` to `options`.

### CODE REVIEW

**Comment by jeffposnick** [comment]
> It's too bad that destructured params increase the file size so much...

I guess that could be why Matt wanted us to avoid them in the core libraries 😄 

This might be the case of something that gets transpiled into bloated code when we need to support older browsers, and once we change our `@babel/preset-env` config in an upcoming v4 release, we'll stop paying the penalty? If you did a comparison of the bundle size in this PR for the current `@babel/preset-env` config and then again with a more recent browser target, what does that size delta look like?

File: packages/workbox-core/_private/cacheWrapper.mjs [hunk 1]
```diff
@@ -29,15 +29,23 @@ import '../_version.mjs';
...
- * @param {string} cacheName
- * @param {Request} request
- * @param {Response} response
- * @param {Array<Object>} [plugins]
+ * @param {Object} options
+ * @param {string} options.cacheName
+ * @param {Request} options.request
+ * @param {Response} options.response
+ * @param {Event} [options.event]
+ * @param {Array} [options.plugins=[]]
```
[the remark this answers is not in the document]
Comment (philipwalton) [reply packages/workbox-core/_private/cacheWrapper.mjs]: Done.

File: packages/workbox-core/_private/fetchWrapper.mjs [hunk 2]
```diff
@@ -27,24 +27,27 @@ import '../_version.mjs';
...
-  // We *should* be able to call `await preloadResponse` even if it's undefined,
-  // but for some reason, doing so leads to errors in our Node unit tests.
-  // To work around that, explicitly check preloadResponse's value first.
-  if (preloadResponse) {
-    const possiblePreloadResponse = await preloadResponse;
+const wrappedFetch = async ({
+    request,
+    event,
+    fetchOptions,
+    preloadResponse,
```
[the remark this answers is not in the document]
Comment (philipwalton) [reply packages/workbox-core/_private/fetchWrapper.mjs]: Done.

**Comment by jeffposnick** [comment]
> > There are also some changes that will need to made to https://developers.google.com/web/tools/workbox/guides/using-plugins#custom_plugins in a separate PR to reflect the new plugin parameters.

> Yes, but I'm assuming that should wait until the next release, right?

Right; we don't have a great way of keeping track of that, so I tend to try to file a PR (or at least create a local commit with the change) against WebFundamentals at around the same time that the Workbox change goes in, and time the merge for when we cut the new Workbox release.

**Comment by jeffposnick** [comment]
I'm curious to see whether the size differential goes down depending on the `@babel/preset-env` target, but I don't think the delta in sizes is blocking.

**Comment by philipwalton** [comment]
I don't think `preset-env` is creating any extra size differential specific to these changes, but it does create a size differential in general (greater than these changes). For example, here's what the gzipped sizes look like comparing master, this PR, and then this PR without any babel transforms:

workbox-core | size
-------------- | ----
master | 2.84 kB
PR | 2.9 kB
PR (w/o babel) | 2.76 kB

workbox-strategies | size
------------------- | -----
master | 1.17 kB
PR | 1.24 kB
PR (w/o babel) | 1.12 kB

AFAICT, the main cause of the size increase is previously we could minify code like this:

```js
cacheWrapper.match(s.e,r,s.s,s.t)
```

But now we're forced to minify it like this (from an actual diff):

```js
cacheWrapper.match({cacheName:r.e,request:s,event:e,matchOptions:r.r,plugins:r.t});
```

And enough of these occurrences adds up I guess...

**Comment by jeffposnick** [comment]
Ah, that makes sense regarding the parameter name mangling. Thanks for checking!

I'm cool with the extra bytes, especially since we have a plan for reducing bundle sizes once we hit v4 and can cut out some of the other transpilation overhead.

**Pull request merged**